File: scWDAC_MATLAB/functions/construct_knnGraph.m
function [H] = construct_knnGraph(X, k, t)
% X : n*d
% t RBF中超参
% k
% 构建RBF相似度矩阵
[n,dim] = size(X);
Dis =EuDist2(X, X);
t = sum(sum(Dis))/n/n;
% A = exp(-Dis/(2*t^2));

[dumb, idx] = sort(Dis, 1);
A_1 = zeros(n);
A_1 = A_1 + diag(ones(n,1));
for i = 1:n
    id = idx(2:k+1,i);
    di = Dis(id, i);
    A_1(id,i) = exp(-di/(2*t^2));
end
% H = exp(-W/(2*t^2));
H = A_1;
H = H-eye(n);
H = H';
% W_e = sum(H);
% % 构建d_v 点的度
% for i = 1:n
%     temp = 0;
%     for j = 1:n
%         if H(i,j) ~= 0
%             temp = temp + H(i,j)*W_e(j);
%         end
%     end
%     d_v(i) = temp; 
% end
% 
% d_e = sum(H);
% 
% D_v = diag(d_v');
% D_e = diag(d_e');
% W = diag(W_e');
% I = diag(ones(n,1));
% L = I - D_v^(-0.5)*H*W*D_e^(-1)*H'*D_v^(-0.5);

end

File: scWDAC_Python/functions/construct_knnGraph.py
from functions.EuDist2 import EuDist2


def construct_knnGraph(X, k, t=None):

    import numpy as np

    n, dim = X.shape
    Dis = EuDist2(X, X)
    t = (np.sum(Dis) / n / n) if (t is None) else t
    idx = np.argsort(Dis, axis=0)
    A_1 = np.zeros((n, n))
    A_1 = A_1 + np.diag(np.ones((n,)))
    for i in range(n):
        id = idx[1:k + 1, i]
        di = Dis[id, i]
        A_1[id, i] = np.exp(-di / (2 * t ** 2))

    H = A_1
    H = H - np.eye(n)
    H = H.T

    return H
